simulator/modules/execute/execute.cpp

```diff
@@ -123,7 +124,10 @@ void Execute<ISA>::clock( Cycle cycle)
         wp_bypass->write( instr.get_dst_v(), cycle);
 
         if ( instr.is_mem_stage_required())
+	{
             wp_mem_datapath->write( std::move( instr), cycle);
```

Comment:
By the way, you may steer here — send jumps to `branch` port, and send loads and stores to `memory` port.

Reply:
I think you can do this:
```c++
if ( instr.is_jump())
wp_prediction_datapath->write( std::move( instr) ,cycle);
else if ( instr.is_mem_stage_required())
    wp_mem_datapath->write( std::move( instr), cycle);
else
    wp_writeback_datapath->write( std::move( instr) , cycle)
```

Reply:
I get this idea, thank you

Reply:
I think, that it should be like this:
```
if ( instr.is_jump())
    wp_branch_datapath->write( std::move( instr) ,cycle);
if ( instr.is_mem_stage_required())
    wp_mem_datapath->write( std::move( instr), cycle);
else
    wp_writeback_datapath->write( std::move( instr) , cycle)
```
because if branch was predicted correct, mem should be writing

Reply:
I think it contains a bug. At least, you cannot do `std::move` here and then you use the instruction — that's not allowed by C++.
(If you wish, you may remove all `std::move`, I'll place them later myself, it's quite complicated language feature).

>  if branch was predicted correct, mem should be writing

Not sure I understand you. What should Mem write?

Reply:
Good night.
> Not sure I understand you. What should Mem write?

I was wrong, jump doesn't load any memory. 
I have found information about `std::move( instr)` and it is really bug and must be solved.
I try to insert this variant
``` c++
if ( instr.is_jump())
wp_prediction_datapath->write( std::move( instr) ,cycle);
else if ( instr.is_mem_stage_required())
    wp_mem_datapath->write( std::move( instr), cycle);
else
    wp_writeback_datapath->write( std::move( instr) , cycle)
```
and unfortunately it always cause this
``` shell
Checker mismatch:
	Checker output: {3}	0x4000fc: bne $t0, $zero, 46
PerfSim output: {4}	0x4001b8: lui $s0, 0x41	 [ $s0 = 0x410000 ]
```
I have found, that problem is in writeback.cpp module:
``` c++
    /* perform checks */
    checker.check( instr);

    /* update simulator cycles info */
    ++executed_instrs;
    last_writeback_cycle = cycle;
    if ( executed_instrs >= instrs_to_run || instr.is_halt())
        wp_halt->write( true, cycle);
```
This functions must be performed for jump (at least it doesn't work without).
Moreover, if I remove rest functions (writers to ports), it decrease IPC.
So, I think that both modules `MEM` and `WRITEBACK` or `BRANCH` and `WRITEBACK` must be performed.

I see one easy way of solving this problem: I need to add `BRANCH_2_WRITEBACK`  datapath port for they communication. So I can keep good variant of steering and execution of both mobules.
I will try to push this variant in hour.

But if it is still is not, what I need to do, I will be pleased to talk with you at closest seminar, if it is possible.
 
Thank you

Reply:
> Also, I can make EXECUTE_2_ port with 3 readers: MEM , BRANCH and WRITEBACK. I think this variant is more logic.

Looks right — I thought you done that already.

> But if it is still is not, what I need to do, I will be pleased to talk with you at closest seminar, if it is possible.

I am on business trip in Haifa, so I will not be able to deliver the last seminar, Kirill will do it.
You may talk to Oleg about branch prediction issues.